### PR TITLE
Add math builtins.

### PR DESCRIPTION
**Pull request by thschroeter**
Work in progress towards support for more math builtins. So far, I've implemented `abs`, `pi` and `sin`. I hope this is what you have in mind. Will add more functions tomorrow. Probably with a `floatBuiltin1` helper.
#2042

### CODE REVIEW

File: sql/parser/builtins.go [hunk 1]
```diff
@@ -443,6 +442,42 @@ var builtins = map[string][]builtin{
...
+			types:      typeList{intType},
+			fn: func(args DTuple) (Datum, error) {
+				return abs(args[0].(DInt))
+			},
+		},
+	},
+	"pi": {
+		builtin{
+			returnType: DummyFloat,
+			types:      nil,
```
Comment (tamird) [inline sql/parser/builtins.go]: this should be `typeList{}`; `nil` means it's variadic

Comment (tamird) [inline sql/parser/builtins.go]: (can you add a comment to the `builtin` struct to explain that? probably would have saved you a minute :)

Comment (thschroeter) [inline sql/parser/builtins.go]: thanks, will do.

File: sql/parser/builtins.go [hunk 2]
```diff
@@ -575,3 +610,13 @@ func datumToRawString(datum Datum) (string, error) {
 
 	return "", fmt.Errorf("argument type unsupported: %s", datum.Type())
 }
+
+func abs(x DInt) (DInt, error) {
+	switch {
+	case x == math.MinInt64:
+		return x, errAbsOfMinInt64
```
Comment (tamird) [inline sql/parser/builtins.go]: just thinking out loud: we can probably support this if we introduce `DUint`. Not sure it's worth it.

We should at least give a better error message, which doesn't mention a go primitive type verbatim.

File: sql/parser/builtins.go [hunk 3]
```diff
@@ -575,3 +610,13 @@ func datumToRawString(datum Datum) (string, error) {
 
 	return "", fmt.Errorf("argument type unsupported: %s", datum.Type())
 }
+
+func abs(x DInt) (DInt, error) {
```
Comment (tamird) [inline sql/parser/builtins.go]: why not inline this into the `builtin`?

**Comment by tamird** [comment]
LGTM so far. Thanks for tackling this, let me know when you're ready for another review.

File: sql/parser/builtins.go (hunk 2 above)
Comment (petermattis) [inline sql/parser/builtins.go]: I don't think we should introduce `DUint` as that would bring along with it all sorts of complexity in handling unsigned vs signed operations. Agreed that the error message should not mention a Go type. Perhaps something like: `abs of min integer value (-9223372036854775808) not defined`

File: sql/testdata/builtin_function [hunk 4]
```diff
@@ -332,3 +332,56 @@ SELECT extract(nanosecond from timestamp '2001-04-10 12:04:59.34565423')
...
+----
+0
+
+query T
+SELECT abs(1.0)
+----
+1
+
+query error abs: abs of MinInt64 is not defined
+SELECT abs(-9223372036854775808)
```
Comment (petermattis) [inline sql/testdata/builtin_function]: Might be good to test `-9223372036854775807` (the smallest integer for which `abs` is defined).

File: sql/testdata/builtin_function [hunk 5]
```diff
@@ -332,3 +332,56 @@ SELECT extract(nanosecond from timestamp '2001-04-10 12:04:59.34565423')
...
+----
+0
+
+query T
+SELECT sin(1.0)
+----
+0.8414709848078965
+
+query T
+SELECT sin(1::float)
```
Comment (petermattis) [inline sql/testdata/builtin_function]: This is going to be equivalent to the previous test as long as casting works (which is tested elsewhere). Perhaps replace with a test for `sin(pi())`.

**Comment by petermattis** [comment]
LGTM. Fleshing out the remaining math functions should be straightforward.

File: sql/parser/builtins.go (hunk 1 above)
Comment (thschroeter) [inline sql/parser/builtins.go]: done

File: sql/parser/builtins.go (hunk 2 above)
Comment (thschroeter) [inline sql/parser/builtins.go]: done

File: sql/parser/builtins.go (hunk 3 above)
Comment (thschroeter) [inline sql/parser/builtins.go]: done

File: sql/testdata/builtin_function (hunk 4 above)
Comment (thschroeter) [inline sql/testdata/builtin_function]: done

File: sql/testdata/builtin_function (hunk 5 above)
Comment (thschroeter) [inline sql/testdata/builtin_function]: Replaced with `abs(sin(pi())) < 1e-12 == true`, let me know if I should kill the new version of the test, if it's too dense for example.

Wrote it like this, since the numeric value of `sin(pi())` is very close to zero, and the implementation of `sin` is platform dependent. Therefore I'm expecting slightly different results for `sin(pi())`, depending on the platform.

**Comment by thschroeter** [comment]
For now I left `round` as a todo. I'm giving it a try tomorrow. So far, I just peeked at the [double_round](https://github.com/python/cpython/blob/master/Objects/floatobject.c#L931) implementation of CPython.

Some functions are only implemented for floats, maybe I can introduce int implementations as well. E.g. for `mod` (with zero division error) and `pow` (with overflow error).

**Comment by thschroeter** [comment]
PTAL, thanks!

File: sql/parser/builtins.go [hunk 6]
```diff
@@ -443,6 +422,171 @@ var builtins = map[string][]builtin{
...
+			},
+		},
+		builtin{
+			returnType: DummyInt,
+			types:      typeList{intType},
+			fn: func(args DTuple) (Datum, error) {
+				x := args[0].(DInt)
+				switch {
+				case x == math.MinInt64:
+					return x, errAbsOfMinInt64
```
Comment (thschroeter) [inline sql/parser/builtins.go]: It's better to return `DNull, errAbsOfMinInt64`.

File: sql/parser/builtins.go [hunk 7]
```diff
@@ -443,6 +422,171 @@ var builtins = map[string][]builtin{
...
+
+	"exp": {
+		floatBuiltin1(func(x float64) (Datum, error) {
+			return DFloat(math.Exp(x)), nil
+		}),
+	},
+
+	"floor": {
+		floatBuiltin1(func(x float64) (Datum, error) {
+			return DFloat(math.Floor(x)), nil
```
Comment (bdarnell) [inline sql/parser/builtins.go]: Should `floor()` (and `ceil`) return a float or an int? I see that Go's `math.Floor()` returns a float, but that's surprising to me.

File: sql/parser/builtins.go [hunk 8]
```diff
@@ -443,6 +422,171 @@ var builtins = map[string][]builtin{
...
+
+	"log": {
+		floatBuiltin1(func(x float64) (Datum, error) {
+			return DFloat(math.Log10(x)), nil
+		}),
+	},
+
+	"mod": {
+		floatBuiltin2(func(x, y float64) (Datum, error) {
+			return DFloat(math.Mod(x, y)), nil
```
Comment (bdarnell) [inline sql/parser/builtins.go]: Should `mod()` have an integer version as well? Every time I've wanted to use `mod` it's been with integers.

File: sql/parser/builtins.go (hunk 7 above)
Comment (petermattis) [inline sql/parser/builtins.go]: They should return a float. Not sure why that is surprising to you. The C-library functions return the same type as the argument.

Comment (bdarnell) [inline sql/parser/builtins.go]: It's surprising because I see these functions as a kind of cast: I want an integer, and I'm specifying how I want that integer created. `floor(x)::int` feels redundant, but I guess there's plenty of precedent for doing it this way (Python's `math.floor()` and friends do return integers)

File: sql/parser/builtins.go (hunk 8 above)
Comment (thschroeter) [inline sql/parser/builtins.go]: Good idea, I've added and integer version of `mod`.

**Comment by thschroeter** [comment]
Added `mod` for integers and a decimal `round` implementation for floats. Not sure if we should support `round` at this point. See discussion in a todo comment in the code.

I'm working on a `pow` implementation for integers with overflow check using Go's `math/big` package.

PTAL, thanks.

File: sql/testdata/builtin_function [hunk 9]
```diff
@@ -332,3 +332,286 @@ SELECT extract(nanosecond from timestamp '2001-04-10 12:04:59.34565423')
...
+----
+1
+
+query T
+SELECT abs(-0.0)
+----
+0
+
+query T
+SELECT abs(1.0)
```
Comment (vivekmenezes) [inline sql/testdata/builtin_function]: Can you add a float that is != to an int.

Also might be worth consolidating some of these tests to look like

SELECT abs(-1.0), abs(1.0), abs(-0.0), abs(0),  abs(1)

File: sql/testdata/builtin_function [hunk 10]
```diff
@@ -332,3 +332,286 @@ SELECT extract(nanosecond from timestamp '2001-04-10 12:04:59.34565423')
...
+
+query T
+SELECT trunc(0.0)
+----
+0
+
+query T
+SELECT trunc(-0.0)
+----
+-0
```
Comment (vivekmenezes) [inline sql/testdata/builtin_function]: hmmm -0 is interesting

File: sql/testdata/builtin_function [hunk 11]
```diff
@@ -332,3 +332,286 @@ SELECT extract(nanosecond from timestamp '2001-04-10 12:04:59.34565423')
...
+
+query T
+SELECT trunc(1.9)
```
Comment (vivekmenezes) [inline sql/testdata/builtin_function]: how about trunc(-1.9) as well

File: sql/testdata/builtin_function [hunk 12]
```diff
@@ -332,3 +332,286 @@ SELECT extract(nanosecond from timestamp '2001-04-10 12:04:59.34565423')
...
+----
+1
+
+query T
+SELECT sin(0.0)
+----
+0
+
+query T
+SELECT sin(1.0)
```
Comment (vivekmenezes) [inline sql/testdata/builtin_function]: Can you add in some negative parameters for all the builtins?

**Comment by vivekmenezes** [comment]
LGTM. Thanks for making this change!

File: sql/testdata/builtin_function (hunk 11 above)
Comment (petermattis) [inline sql/testdata/builtin_function]: At some point we're just testing the behavior of the routines we're calling in the `math` package, which isn't terribly interesting. I think a single test case per input type and boundary condition in our implementation (not the `math` implementation) is sufficient.

File: sql/parser/builtins.go [hunk 13]
```diff
@@ -443,6 +425,219 @@ var builtins = map[string][]builtin{
...
+	"pow":   {powImpl},
+	"power": {powImpl},
+
+	"radians": {
+		floatBuiltin1(func(x float64) (Datum, error) {
+			return DFloat(x * math.Pi / 180.0), nil
+		}),
+	},
+
+	// TODO(thschroeter): should we support round for our floatType?
```
Comment (petermattis) [inline sql/parser/builtins.go]: MySQL supports `round` for `int`, `float` and `numeric`. I think it makes sense to support for `int` and `float`, but I'm not thrilled with the implementation below going through `Sprintf` and `ParseFloat`.

Comment (petermattis) [inline sql/parser/builtins.go]: I'd suggest removing this implementation from this PR and adding a TODO to implement in a follow-up PR. Unlike many of the other functions you've implemented, this one is tricky.

File: sql/parser/builtins.go [hunk 14]
```diff
@@ -443,6 +425,219 @@ var builtins = map[string][]builtin{
...
+			return DFloat(math.Sin(x)), nil
+		}),
+	},
+
+	"sign": {
+		floatBuiltin1(func(x float64) (Datum, error) {
+			if math.Signbit(x) {
+				return DFloat(-1.0), nil
+			}
+			return DFloat(1), nil
```
Comment (petermattis) [inline sql/parser/builtins.go]: The postgres docs say `sign(0.0) = 0.0`.

File: sql/parser/builtins.go [hunk 15]
```diff
@@ -443,6 +425,219 @@ var builtins = map[string][]builtin{
...
+		}),
+		builtin{
+			returnType: DummyInt,
+			types:      typeList{intType},
+			fn: func(args DTuple) (Datum, error) {
+				if args[0].(DInt) < 0 {
+					return DInt(-1), nil
+				}
+				return DInt(1), nil
```
Comment (petermattis) [inline sql/parser/builtins.go]: Similarly, `sign(0) = 0`.

File: sql/parser/builtins.go [hunk 16]
```diff
@@ -443,6 +425,219 @@ var builtins = map[string][]builtin{
...
+		builtin{
+			returnType: DummyFloat,
+			types:      typeList{floatType, intType},
+			fn: func(args DTuple) (Datum, error) {
+				n := args[1].(DInt)
+				if n < 0 {
+					return DNull, errRoundNegativeDigits
+				}
+				x := args[0].(DFloat)
+				f := fmt.Sprintf("%%.%df", n)
```
Comment (bdarnell) [inline sql/parser/builtins.go]: This could use `strconv.FormatFloat(x, 'f', n, 64)` instead of the double `Sprintf`(if it doesn't get removed completely). We also shouldn't ignore the error from `ParseFloat` below even though it really shouldn't be possible.

File: sql/parser/builtins.go [hunk 17]
```diff
@@ -443,6 +425,219 @@ var builtins = map[string][]builtin{
...
+	// is a base-10k floating point decimal, where decimal rounding
+	// is more efficient compared to our base-2 float.
+	//
+	// The naive implementation below is not efficient since it's
+	// going through an intermediate string. It seems possible
+	// to provide a more efficient implementation that operates
+	// floats without allocating an intermediate string to represent
+	// the number as decimal before rounding. Do we need that?
+	// How is round used? In case it's mostly used to present a
+	// handful of results in a readable way it doesn't need to be
```
Comment (bdarnell) [inline sql/parser/builtins.go]: Rounding for presentation purposes in a base-2 float is tricky - rounding errors may creep back in depending on what client is being used. For example, in python 2.6:
```
>>> round(math.pi, 5) 
3.1415899999999999
```

In practice, many uses of `round()` may be better served by a function that takes a float and a precision and returns a string (although that function should not be called `round()`, since people would expect that function to return a number).

File: sql/testdata/builtin_function (hunk 9 above)
Comment (thschroeter) [inline sql/testdata/builtin_function]: Done. Grouping test cases helped me finding and removing some duplicates.

File: sql/testdata/builtin_function (hunk 12 above)
Comment (thschroeter) [inline sql/testdata/builtin_function]: done

File: sql/parser/builtins.go (hunk 15 above)
Comment (thschroeter) [inline sql/parser/builtins.go]: done

File: sql/parser/builtins.go (hunk 14 above)
Comment (thschroeter) [inline sql/parser/builtins.go]: done

File: sql/parser/builtins.go (hunk 13 above)
Comment (thschroeter) [inline sql/parser/builtins.go]: done

File: sql/parser/builtins.go [hunk 18]
```diff
@@ -443,6 +424,190 @@ var builtins = map[string][]builtin{
...
+
+	"sign": {
+		floatBuiltin1(func(x float64) (Datum, error) {
+			switch {
+			case x < 0:
+				return DFloat(-1), nil
+			case x == 0:
+				return DFloat(0), nil
+			}
+			return DFloat(1), nil
```
Comment (petermattis) [inline sql/parser/builtins.go]: I wonder if you should continue to use `math.Signbit` here. Note that `-0.0 == 0.0`, but `math.Signbit(-0.0) != math.Signbit(0.0)`.

Comment (petermattis) [inline sql/parser/builtins.go]: Oh, I see this is the postgres behavior.

File: sql/parser/builtins.go [hunk 19]
```diff
@@ -31,10 +32,15 @@ import (
 )
 
 var errEmptyInputString = errors.New("the input string should not be empty")
+var errAbsOfMinInt64 = errors.New("abs of min integer value (-9223372036854775808) not defined")
+var errDivisionByZero = errors.New("division by zero")
+var errRoundNegativeDigits = errors.New("number of digits should be greater or equal 0")
```
Comment (petermattis) [inline sql/parser/builtins.go]: This is unused.

File: sql/parser/builtins.go [hunk 20]
```diff
@@ -443,6 +424,190 @@ var builtins = map[string][]builtin{
...
+		builtin{
+			returnType: DummyInt,
+			types:      typeList{intType, intType},
+			fn: func(args DTuple) (Datum, error) {
+				y := args[1].(DInt)
+				if y == 0 {
+					return DNull, errDivisionByZero
+				}
+				x := args[0].(DInt)
+				return DInt(x - (x/y)*y), nil
```
Comment (petermattis) [inline sql/parser/builtins.go]: Is there a reason you're not doing `x % y` here?

File: sql/parser/builtins.go [hunk 21]
```diff
@@ -31,10 +32,15 @@ import (
 )
 
 var errEmptyInputString = errors.New("the input string should not be empty")
+var errAbsOfMinInt64 = errors.New("abs of min integer value (-9223372036854775808) not defined")
+var errDivisionByZero = errors.New("division by zero")
```
Comment (petermattis) [inline sql/parser/builtins.go]: This will be unused when you switch to using `errZeroModulus`.

**Comment by petermattis** [comment]
LGTM modulo small tweaks to integer mod.

File: sql/parser/builtins.go (hunk 18 above)
Comment (thschroeter) [inline sql/parser/builtins.go]: Yes, I thought it's probably best to follow postgres behavior if it's easy to do. Before I was sticking closer to the math functions Go provides. I guess, it would be good to compare the results of the math builtins against postgres and document differences.

File: sql/parser/builtins.go (hunk 20 above)
Comment (thschroeter) [inline sql/parser/builtins.go]: I thought Go was so streamlined that `%` doesn't exist, obviously I was wrong.

**Comment by thschroeter** [comment]
Can I squash, rebase and merge?

File: sql/testdata/builtin_function [hunk 22]
```diff
@@ -332,3 +332,178 @@ SELECT extract(nanosecond from timestamp '2001-04-10 12:04:59.34565423')
...
+SELECT sqrt(-1.0), sqrt(4.0)
+----
+NaN 2
+
+query T
+SELECT tan(-5.0), tan(0.0), tan(5.0)
+----
+3.3805150062465854 0 -3.3805150062465854
+
+
```
Comment (petermattis) [inline sql/testdata/builtin_function]: Spurious extra blank line here.

File: sql/parser/builtins.go [hunk 23]
```diff
@@ -61,32 +65,7 @@ func (b builtin) match(types typeList) bool {
...
-//   - pi
-//   - pow/power
-//   - radians
-//   - round
-//   - sign
-//   - sin
-//   - sqrt
-//   - tan
-//   - trunc
+// TODO(pmattis): Need to flesh out the supported functions.
```
Comment (petermattis) [inline sql/parser/builtins.go]: I think we can just remove this TODO completely now.

**Comment by petermattis** [comment]
LGTM. Squash, rebase and merge.

**Comment by thschroeter** [comment]
Thanks for your review!
